### PR TITLE
V8: Don't retain "cculture" when navigating away from the current content

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes: https://github.com/umbraco/Umbraco-CMS/issues/4468

### Description

See #4468 for description and steps to reproduce.

This PR ensures that we *don't* retain `cculture` when navigating away from the current content. It looks like this:

dont-retain-cculture-after

I have tested that it works with (doesn't break any of) the current other usages of `clearSearch()` - i.e. when creating new content or navigating between sections.

### CODE REVIEW

**Comment by bergmania** [comment]
We introduced it, to store the `cculture`, when we used the breadcrumb. https://github.com/umbraco/Umbraco-CMS/issues/3648

**Comment by kjac** [comment]
But there's no guarantee that the parent content in the breadcrumb path is culture variant.

That aside there must be a better way to handle the breadcrumb than to retain cculture everywhere..?

**Comment by clausjensen** [comment]
@kjac I've just gone over this with @bergmania .. we're both confused .. but think we might have a solution 😃 
If we use your fix ... we basically break the breadcrumb stuff. If we don't use your fix - the switching language part from the tree is broken.

So .. what we came up with is that we _think_ that it could work if we keep the retained cculture as it was before, while also applying this change:

- We know the tree always reflects the mculture.
- When selecting a node in the tree, you would always expect to have the node open up in the language of what you just clicked on - so if the mculture (and names shown in the tree) are set to the english ones - you would expect that when clicking one of them - it would open up the content editing in that culture.
- We think this should be achievable by simply setting the cculture to what the mculture is whenever you click a node in the tree.

What do you think - did we miss anything?

**Comment by kjac** [comment]
@clausjensen I think you're venturing onto dangerous grounds here 🤔 

First and foremost I don't see any difference between navigating using the tree or using the breadcrumb. I'd expect the cculture selection to be a one-off that's forgotten when I move away from content I'm currently looking at, since I have chosen my main (default) culture to be just that.

That being said there may be different opinions when it comes specifically to the breadcrumb. Fair enough 😄 but then we should explicitly retain cculture when operating the breadcrumb, not as a general rule. And all the necessary handles for doing just that are already in the codebase - I'd be happy to update this PR accordingly.

Thoughts?

**Comment by bergmania** [comment]
From my point of view. The most important thing is that, the link is to the same culture as the name indicated. Breadcrumbs uses the culture specific name. Main tree items don't.

**Comment by kjac** [comment]
Point well made. Want me to update the PR to retain cculture in the breadcrumb, or would you rather have me hack the tree to forcefully remove cculture there? 

Warning: the latter option is not supported as of yet and might complicate the navigation service a bit.

**Comment by clausjensen** [comment]
Haha - yes! .. very dangerous grounds especially for backend developers to venture into! 😃 

Here's my thoughts:
- When looking at the breadcrumb - it is in context of the content (window) .. so I would expect it to be localized to what my content window is set to. I therefore also expect that when I use it to navigate - I would navigate to the language of what I'm clicking on. If I change the language of my content window - I'd expect the breadcrumb to change language as well... and that the links get me to what they "indicate" language-wise.

- Pretty much the same goes for the tree. It is a separate context here (in my opinion) - so when I change main culture and my tree now shows danish nodes. Clicking one of them would navigate me to that node - editing it in danish.

I think solution wise - I'd prefer if we could keep the `var retainedQueryStrings = ["mculture", "cculture"];` .. as I really don't know if there's other things apart from the breadcrumb that we haven't noticed being affected if we change that to not retain cculture. Do you think it's possible to hack the tree/service in a not-so-bad way? 😄

**Comment by kjac** [comment]
@clausjensen I'll have a look 👍

**Comment by kjac** [comment]
@clausjensen @bergmania as it turns out the "retained querystrings" are explicitly and forcefully appended on every single route change - well outside the scope of the tree.

This leaves us with no other option than to retain cculture explicitly when needed, and leave it out of the generally "retained querystrings". Sorry.

**Comment by clausjensen** [comment]
Crap 😃 
Well that's just how it's going to be then. I guess it _is_ also the nicer solution... and we'll just have to hope there isn't any sideeffects we haven't thought of

**Comment by Shazwazza** [comment]
I have an idea and its based on component inheritance for the breadcrumbs
and doesn't rely on query strings, I can explain more when I wake up and
see if that will work better.

**Comment by kjac** [comment]
Cool @Shazwazza... sleep sweet 😴 

In the mean time I have updated the PR to retain cculture in the breadcrumb. It got a bit out of of hand because I also wanted to retain support for opening the breadcrumb items in new tabs/windows... 😆

**Comment by Shazwazza** [comment]
So the `<umb-breadcrumbs>` directive is used in the content editor directive like:

```html
<umb-breadcrumbs
    ng-if="ancestors && ancestors.length > 0"
    ancestors="ancestors"
    entity-type="content">
</umb-breadcrumbs>
```

We can easily pass in the current culture to this directive explicilty, not worrying about any query string, the culture is already available as a property so can just be:

```html
<umb-breadcrumbs
    ng-if="ancestors && ancestors.length > 0"
    ancestors="ancestors"
    entity-type="content"
    culture="culture">
</umb-breadcrumbs>
```

Then it's also used in `<umb-mini-list-view>` which is used in various pickers. 

Since the breadcrumb is a component (directive), component's need to be created with as little external/magic dependencies as possible, just like any object. If the breadcrumb needs to know about culture, it needs to be told about it, not by just fetching it out of thin air from the query string so please pass this in as a property. Then this becomes flexible and its up to the implementer of this component to determine the culture. This can be done in many ways, not just the query string. For example, in split view, we don't use the query string, each editor knows about it's own culture and in fact if you have a component within the editor, you can require "^^umbVariantContent" to access that component and extract the current model and culture from it. That component contains everything about what is rendering in the content editor for each individual split view.

So my point is, we should be able to access everything we need and pass along any culture information downwards to where it needs to go and not rely on global state such as query strings - those should be used only at the highest levels.

File: src/Umbraco.Web.UI.Client/src/common/directives/components/editor/umbbreadcrumbs.directive.js [hunk 1]
```diff
@@ -58,6 +58,28 @@ Use this directive to generate a list of breadcrumbs.
...
+                ) {
+                    // yes, let the link open itself
+                    return;
+                }
+                event.stopPropagation();
+                event.preventDefault();
+
+                var path = scope.pathTo(ancestor);
+                $location.path(path);
+                navigationService.clearSearch(["cculture"]);
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/common/directives/components/editor/umbbreadcrumbs.directive.js]: @kjac i think i'm confused but need to double check here, i though the purpose of this PR is to not automatically retain cculture but for the breadcrumb links we would retain it but this seems to be explicitly clearing it?

Comment (kjac) [reply src/Umbraco.Web.UI.Client/src/common/directives/components/editor/umbbreadcrumbs.directive.js]: Hehe it does look weird, but the input array contains the parameters that should *not* be cleared from search. Yeah...

Comment (Shazwazza) [reply src/Umbraco.Web.UI.Client/src/common/directives/components/editor/umbbreadcrumbs.directive.js]: ahh, haha... i should read my own APIs, lol, that clears all except for that :) all good

Comment (kjac) [reply src/Umbraco.Web.UI.Client/src/common/directives/components/editor/umbbreadcrumbs.directive.js]: Basically clearSearch clears everything but the passed keys.

**Pull request merged**